hudi-client/hudi-client-common/src/main/java/org/apache/hudi/client/AbstractHoodieClient.java

```diff
@@ -124,9 +128,30 @@ public HoodieWriteConfig getConfig() {
     return config;
   }
 
+  private void checkIndexTypeCompatible(String writeIndexType, String persistIndexType) {
+    if (writeIndexType.equals(persistIndexType)) {
+      return;
+    } else {
+      if ((persistIndexType.equals(HoodieIndex.IndexType.GLOBAL_BLOOM.name())
+          && (writeIndexType.equals(HoodieIndex.IndexType.BLOOM.name()) || writeIndexType.equals(HoodieIndex.IndexType.SIMPLE.name())))
+          || writeIndexType.equals(HoodieIndex.IndexType.INMEMORY.name())) {
+        return;
+      } else {
+        throw new HoodieException("The new write indextype " + writeIndexType
```

Comment:
One tricky thing here is  : we have made HoodieIndex pluggable cc @leesf . So, the index type there would be unknown? How about adding a CUSTOM index type and a new `indexType()` to HoodieIndex class? then we can treat that here as something that is not compatible with any other index type. 

Right now, we write null for this case? So this method could error in that scenario. Just saying we need to think about this.

Reply:
Understand what you mean. it is so helpful
1. As index is pluggable. In class SparkHoodieIndex will first construct index using hoodie.index.class, at this time the indextype will be null.
2. I think should to add indexType() to HoodieIndex class. And the index type should get from  indexType() .
3. also  the  checkIndexTypeCompatible can in SparkHoodieIndex.createIndex
4. the index type persist to properties file ,should get from SparkHoodieIndex.createIndex().indexType()

also other pluggable with more than one parameter control  implement should do like this  to get the implement type.

Reply:
Not following the last sentence. but high level, we are on the same page I think. I was suggesting `CUSTOM` as the indexType() when its user supplied class. we can make the compatibility check such that custom is not compatible with anything else. 

I think you are suggesting to persist the index class as well?

Reply:
@vinothchandar   i am doing persist the hoodieindex.indexType(). get indexType will first instantiate hoodieIndex,then get the indexType(). what do you think about this choice?